data/src/main/java/org/apache/iceberg/data/PartitionStatsWriterUtil.java

```diff
@@ -0,0 +1,106 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.data;
+
+import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT;
+import static org.apache.iceberg.TableProperties.DEFAULT_FILE_FORMAT_DEFAULT;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Iterator;
+import java.util.List;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.HasTableOperations;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.PartitionStatsUtil;
+import org.apache.iceberg.Partitioning;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.SortOrder;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.data.parquet.GenericParquetReaders;
+import org.apache.iceberg.encryption.EncryptedFiles;
+import org.apache.iceberg.encryption.EncryptionKeyMetadata;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.DataWriter;
+import org.apache.iceberg.io.FileWriterFactory;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.parquet.Parquet;
+import org.apache.iceberg.types.Types;
+
+public final class PartitionStatsWriterUtil {
+
+  private PartitionStatsWriterUtil() {}
+
+  public static OutputFile newPartitionStatsFile(Table table, long snapshotId) {
+    FileFormat fileFormat =
+        FileFormat.fromString(
+            table.properties().getOrDefault(DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT));
+    return table
+        .io()
+        .newOutputFile(
+            ((HasTableOperations) table)
+                .operations()
+                .metadataFileLocation(
+                    fileFormat.addExtension(String.format("partition-stats-%d", snapshotId))));
+  }
+
+  public static void writePartitionStatsFile(
+      Table table, Iterator<Record> records, OutputFile outputFile) {
+    Schema dataSchema = PartitionStatsUtil.schema(Partitioning.partitionType(table));
+    FileFormat fileFormat =
+        FileFormat.fromString(
+            table.properties().getOrDefault(DEFAULT_FILE_FORMAT, DEFAULT_FILE_FORMAT_DEFAULT));
+    FileWriterFactory<Record> factory =
+        GenericFileWriterFactory.builderFor(table)
+            .dataSchema(dataSchema)
+            .dataFileFormat(fileFormat)
+            .dataSortOrder(sortOrder(dataSchema))
```

Comment:
sorting as per the unified partition tuple order as per the spec.

Reply:
I don't think this really does the sorting. Seems it only marks the generated file with the sort ID?

Reply:
Yeah I'm not sure this actually has any use, not sure

Reply:
I see. I assumed it will sort the data as we are passing to writer. Looks like it just adds sort order to metadata. 

During Spec design phase, we decided that it is good to sort the partition stats before writing to a file so that it can help for faster lookup and better compression. But seems like we need engine to sort it. Just the Java API of parquet writers we have doesn't have the option to sort it?

So, we should update back the spec that these fields cannot be sorted in a separate PR. WDYT?

Reply:
How about using a sorted map to store the stats entries? I suppose we can get a comparator for the partition data using the `Comparators.forType` util, right?

Reply:
done